Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp

```diff
@@ -1481,116 +1649,68 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
 {
-    auto accumulateIntrinsicSizes = [&](RenderBox* gridItem) {
-        bool isNewEntry = itemsSet.add(*gridItem).isNewEntry;
-        GridSpan span = m_renderGrid->gridSpanForGridItem(*gridItem, m_direction);
+    auto populateDefiniteItems = [&](unsigned trackIndex, GridSpan& gridSpan, unsigned spanLength, CheckedPtr<RenderBox> gridItem, Vector<GridTrack> & allTracks) {
+        if (gridSpan.startLine() != trackIndex)
+            return;
 
-        // Masonry Track Sizing
-        // https://drafts.csswg.org/css-grid-3/#track-sizing
-        // We should only compute track sizes on a subset of items.
-        //
-        // - Items explicitly placed in that track contribute.
-        // - Items without an explicit placement contribute (regardless of whether they are ultimately placed in that track).
+        auto minContentForGridItem = m_strategy->minContentForGridItem(*gridItem, gridLayoutState);
+        auto maxContentForGridItem = m_strategy->maxContentForGridItem(*gridItem, gridLayoutState);
+        auto minSizeForGridItem = m_strategy->minSizeForGridItem(*gridItem, gridLayoutState);
 
-        // m_direction shall always be the gridAxisDirection.
-        ASSERT(!isDirectionInMasonryDirection());
+        bool spansFlexTracks = spanningItemCrossesFlexibleSizedTracks(gridSpan);
 
-        isNewEntry = true;
+        if (spanLength == 1 && !spansFlexTracks)
+            sizeTrackToFitNonSpanningItem(gridSpan, *gridItem, allTracks[trackIndex], gridLayoutState);
+        else {
+            auto minMaxTrackSizeWithGridSpan = MasonryMinMaxTrackSizeWithGridSpan { MasonryMinMaxTrackSize { minContentForGridItem, maxContentForGridItem, minSizeForGridItem }, gridSpan };
 
-        // Correct the span as the grid item is coming from another track.
-        auto shift = 0u;
-        auto gridItemEndIndex = span.integerSpan() + trackIndex;
-        if (span.integerSpan() > 1 && (gridItemEndIndex > tracks(m_direction).size())) {
-            shift = gridItemEndIndex - tracks(m_direction).size();
-            shift = (shift > trackIndex) ? 0 : shift;
+            if (spansFlexTracks)
+                definiteItemsThatSpanFlexTrack.append(minMaxTrackSizeWithGridSpan);
+            else
+                definiteItems[spanLength].append(minMaxTrackSizeWithGridSpan);
         }
+    };
 
-        span = GridSpan::translatedDefiniteGridSpan(trackIndex - shift, trackIndex + span.integerSpan() - shift);
-
-        if (shouldExcludeGridItemForMasonryTrackSizing(*gridItem, trackIndex, span))
-            return;
-
-        if (CheckedPtr inner = dynamicDowncast<RenderGrid>(gridItem); inner && inner->isSubgridInParentDirection(iterator.direction())) {
-            // Contribute the mbp of wrapper to the first and last tracks that we span.
-            GridSpan subgridSpan = GridSpan::translatedDefiniteGridSpan(trackIndex, trackIndex + span.integerSpan());
+    auto populateIndefiniteItems = [&](CheckedPtr<RenderBox> gridItem, unsigned spanLength) {
```

Comment:
You already protect `gridItem` on the stack so this should not be a `CheckedPtr` and should just be a raw pointer.